### ISSUE TITLE
client side OOME 3.9.4 3.10.1 cach config

### ISSUE THREAD

**Issue by Danny-Hazelcast** [hazelcast/hazelcast#13137]
from https://hazelcast-l337.ci.cloudbees.com/view/rollup/job/rollup-All_3.9.4_3.10.1/25/console
```
16:07:48 /disk1/jenkins/workspace/rollup-All_3.9.4_3.10.1/3.10.1/2018_05_17-12_40_22/stable/destroy/create-use-destroy-HDnearcache Failed
16:07:48 fail HzClient6HZ _create-use-destroy_createUseDistDestroy_cacheBak1HD-ncHD hzcmd.cache.CreateUseDestroy threadId=2 java.lang.Exception: java.lang.OutOfMemoryError: Java heap space 
16:07:48 http://54.82.84.143/~jenkins/workspace/rollup-All_3.9.4_3.10.1/3.10.1/2018_05_17-12_40_22/stable/destroy/create-use-destroy-HDnearcache
```

it looks like the 3.10.1 client hit the same OOME issue as caused by this line - https://github.com/hazelcast/hazelcast/blob/14dac86b78ec5048f16756d0d0bca04c92987801/hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/ClientCacheHelper.java#L101
Where it ends up reading a String and OOME’s when constructing the `charBuffer = new char[charCount];`

```
java.lang.OutOfMemoryError: Java heap space
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataInput.readUTF(ByteArrayObjectDataInput.java:558)
	at com.hazelcast.config.CacheConfig.readData(CacheConfig.java:567)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.readInternal(DataSerializableSerializer.java:158)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:105)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:50)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:48)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:187)
	at com.hazelcast.client.cache.impl.ClientCacheHelper.deserializeCacheConfig(ClientCacheHelper.java:101)
	at com.hazelcast.client.cache.impl.ClientCacheHelper.getCacheConfig(ClientCacheHelper.java:78)
	at com.hazelcast.client.cache.impl.HazelcastClientCacheManager.getCacheConfig(HazelcastClientCacheManager.java:103)
	at com.hazelcast.client.cache.impl.HazelcastClientCacheManager.findCacheConfig(HazelcastClientCacheManager.java:126)
	at com.hazelcast.cache.impl.AbstractHazelcastCacheManager.getCacheUnchecked(AbstractHazelcastCacheManager.java:238)
	at com.hazelcast.cache.impl.AbstractHazelcastCacheManager.getCache(AbstractHazelcastCacheManager.java:220)
	at com.hazelcast.cache.impl.AbstractHazelcastCacheManager.getCache(AbstractHazelcastCacheManager.java:67)
	at hzcmd.cache.CreateUseDestroy.timeStep(CreateUseDestroy.java:11)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)
	at remote.bench.BenchThread.call(BenchThread.java:38)
	at remote.bench.BenchThread.call(BenchThread.java:12)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:473)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
	at java.lang.Thread.run(Thread.java:748)
java.lang.OutOfMemoryError: Java heap space
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataInput.readUTF(ByteArrayObjectDataInput.java:558)
	at com.hazelcast.config.CacheConfig.readData(CacheConfig.java:567)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.readInternal(DataSerializableSerializer.java:158)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:105)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:50)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:48)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:187)
	at com.hazelcast.client.cache.impl.ClientCacheHelper.deserializeCacheConfig(ClientCacheHelper.java:101)
	at com.hazelcast.client.cache.impl.ClientCacheHelper.getCacheConfig(ClientCacheHelper.java:78)
	at com.hazelcast.client.cache.impl.HazelcastClientCacheManager.getCacheConfig(HazelcastClientCacheManager.java:103)
	at com.hazelcast.client.cache.impl.HazelcastClientCacheManager.findCacheConfig(HazelcastClientCacheManager.java:126)
	at com.hazelcast.cache.impl.AbstractHazelcastCacheManager.getCacheUnchecked(AbstractHazelcastCacheManager.java:238)
	at com.hazelcast.cache.impl.AbstractHazelcastCacheManager.getCache(AbstractHazelcastCacheManager.java:220)
	at com.hazelcast.cache.impl.AbstractHazelcastCacheManager.getCache(AbstractHazelcastCacheManager.java:67)
	at hzcmd.cache.CreateUseDestroy.timeStep(CreateUseDestroy.java:11)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)
	at remote.bench.BenchThread.call(BenchThread.java:38)
	at remote.bench.BenchThread.call(BenchThread.java:12)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:473)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
	at java.lang.Thread.run(Thread.java:748)
```

looks like 
https://github.com/mmedenjak/hazelcast/tree/preJoinCacheConfig-3.9.5-3.10-fix

is a fix for this issue

**Comment by Danny-Hazelcast** [comment]
passing tests with https://github.com/mmedenjak/hazelcast/tree/preJoinCacheConfig-3.9.5-3.10-fix

https://hazelcast-l337.ci.cloudbees.com/view/split/job/split-data-loss-multi-v/19/console

https://hazelcast-l337.ci.cloudbees.com/view/rollup/job/rollup-All_3.9.4_3.10.1-client-npe/2/console

**Comment by vbekiaris** [comment]
Fix #13157 is now merged in `maintenance-3.x` branch, closing issue as it has been previously verified that tests pass.
